src/flb_lib.c

```diff
@@ -549,6 +549,18 @@ int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                         void (*out_callback) (void *, int, int, void *, size_t, void *),
                         void *out_callback_data,
                         void *test_ctx)
+{
+    return flb_output_set_test_with_ctx_callback(ctx, ffd, test_name, out_callback,
+                                                 out_callback_data, test_ctx, NULL);
+}
+
+int flb_output_set_test_with_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
+                        void (*out_callback) (void *, int, int, void *, size_t, void *),
+                        void *out_callback_data,
+                        void *test_ctx,
+                        void *(*test_ctx_callback) (struct flb_config *,
+                                                    struct flb_input_instance *,
+                                                    void *, void *))
```

Comment:
This change shouldn't be good for maintanancibility.

Instead, how about using the following style?

```c
int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                        void (*out_callback) (void *, int, int, void *, size_t, void *),
                        void *out_callback_data,
                        void *test_ctx)
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_mode = FLB_TRUE;
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.rt_out_callback = out_callback;
        o_ins->test_formatter.rt_data = out_callback_data;
        o_ins->test_formatter.flush_ctx = test_ctx;
        o_ins->test_formatter.flush_ctx_callback = NULL;
    }
    else {
        return -1;
    }

    return 0;
}

int flb_output_set_test_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
                                     void *(*test_ctx_callback) (struct flb_config *,
                                                                 struct flb_input_instance *,
                                                                 void *, void *))
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.flush_ctx_callback = test_ctx_callback;
    }
    else {
        return -1;
    }

    return 0;
}
```

Plus, the corresponsed C header difinitions like:

```c
FLB_EXPORT int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                                   void (*out_callback) (void *, int, int,
                                                         void *, size_t, void *),
                                   void *out_callback_data,
                                   void *test_ctx);
FLB_EXPORT int flb_output_set_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
                                           void *(*test_ctx_callback) (struct flb_config *,
                                                                       struct flb_input_instance *,
                                                                       void *, void *));
```

Adding simply the new argument is not good idea to maintain code base.

If you use these suggested functions, you also need to change for unit testing file of out_elasticsearch plugin.

Reply:
I'm sorry, but the tests for Elasticsearch output plugin need both callbacks at the same time - not just one of `out_callback` or `test_ctx_callback`. The `out_callback` is used in the tests for validation / assertion, while `test_ctx_callback` is used to extract the right value used later as `out_context` parameter of Elasticsearch output plugin's `cb_es_flush` function (which is `cb_flush` callback of Fluent Bit output plugin). This change is not just about adding one more parameter, but is about providing tests with flexibility to make tested plugin acting just like it does during its work, where the "optional flush context" parameter of formatter function is not a constant but is controlled by `cb_flush` callback.

The same approach could be used for the tests for Forward output plugin, i.e. this change is not just for Elasticsearch plugin. The only reason tests for Forward output plugin doesn't use this approach is because of https://github.com/fluent/fluent-bit/blob/master/plugins/out_forward/forward_format.c#L563 where, IMHO, `flb_forward_format` function breaks its responsibility area just to not break the tests :)

Reply:
... maybe if we change `flb_output_set_test` to something like (removing line `o_ins->test_formatter.flush_ctx_callback = NULL;`):

```cpp
int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                        void (*out_callback) (void *, int, int, void *, size_t, void *),
                        void *out_callback_data,
                        void *test_ctx)
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_mode = FLB_TRUE;
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.rt_out_callback = out_callback;
        o_ins->test_formatter.rt_data = out_callback_data;
        o_ins->test_formatter.flush_ctx = test_ctx;
    }
    else {
        return -1;
    }

    return 0;
}
```

Then we can use `flb_output_set_test` and `flb_output_set_ctx_callback` together (one-by-one) achieving the same as we have right now in this pull request. But due to `test_formatter` is not initialized somewhere else, this approach will require one more function to fill `test_formatter.flush_ctx` and `test_formatter.flush_ctx_callback` with `NULL` which sounds like a complication (will require a lot of changes in existing tests - we will need to call that init function everywhere before `flb_output_set_test` function).

Reply:
Implemented in 7ee017f7080c34ec15e11e06e6cdc556ba55f1f4. I called new function `flb_output_set_test_flush_ctx_callback` instead of `flb_output_set_ctx_callback`. Thank you for a good suggestion!